packages/flutter_goldens/lib/flutter_goldens.dart

```diff
@@ -0,0 +1,202 @@
+    final File goldenFile = _getGoldenFile(golden);
+    await goldenFile.parent.create(recursive: true);
+    await goldenFile.writeAsBytes(imageBytes, flush: true);
+  }
+
+  File _getGoldenFile(Uri uri) {
+    final File relativeFile = fs.file(uri);
+    final Directory testDirectory = fs.directory(basedir);
+    final String relativeBase = fs.path.relative(testDirectory.path, from: goldens.flutterRoot.path);
+    return goldens.repositoryRoot.childDirectory(relativeBase).childFile(relativeFile.path);
+  }
+}
+
+/// A class that represents a clone of the https://github.com/flutter/goldens
+/// repository, nested within the `bin/cache` directory of the caller's Flutter
+/// repository.
+@visibleForTesting
+class GoldensClient {
+  GoldensClient({
+    this.fs: const LocalFileSystem(),
+    this.platform: const LocalPlatform(),
+    this.process: const LocalProcessManager(),
+  });
+
+  final FileSystem fs;
+  final Platform platform;
+  final ProcessManager process;
+
+  RandomAccessFile _lock;
+
+  Directory get flutterRoot => fs.directory(platform.environment[_kFlutterRootKey]);
+
+  Directory get repositoryRoot => flutterRoot.childDirectory(fs.path.join('bin', 'cache', 'pkg', 'goldens'));
+
+  /// Prepares the local clone of the `flutter/goldens` repository for golden
+  /// file testing.
+  ///
+  /// This ensures that the goldens repository has been cloned into its
+  /// expected location within `bin/cache` and that it is synced to the Git
+  /// revision specified in `bin/internal/goldens.version`.
+  ///
+  /// While this is preparing the repository, it obtains a file lock such that
+  /// [GoldensClient] instances in other processes or isolates will not
+  /// duplicate the work that this is doing.
+  Future<void> prepare() async {
+    if (await _isSyncRequired()) {
+      await _obtainLock();
+      try {
+        // Check if sync is required again now that we have the lock.
+        if (await _isSyncRequired()) {
+          await _doSync();
+        }
+      } finally {
+        await _releaseLock();
+      }
+    }
+  }
+
+  Future<bool> _isSyncRequired() async {
+    if (!repositoryRoot.existsSync()) {
+      return true;
+    } else {
+      final io.ProcessResult revParse = await process.run(
+        <String>['git', 'rev-parse', 'HEAD'],
+        workingDirectory: repositoryRoot.path,
+      );
+      if (revParse.exitCode != 0) {
+        return true;
+      }
+      final File versionFile = flutterRoot.childFile(fs.path.join('bin', 'internal', 'goldens.version'));
+      final String currentVersion = revParse.stdout.trim();
+      final String goldensVersion = (await versionFile.readAsString()).trim();
+      return currentVersion != goldensVersion;
+    }
+  }
+
+  Future<void> _doSync() async {
+    await repositoryRoot.create(recursive: true);
+    await _runCommands(
+      <String>[
+        'git init',
+        'git remote add origin https://github.com/flutter/goldens.git',
+        'git pull origin master',
```

Comment:
what'll happen if you have local changes?

Reply:
If there are merge conflicts, it'll throw and fail your test run.

Reply:
(otherwise, it'll just work)